source/common/filter/http/filter_config_discovery_impl.cc

```diff
@@ -169,52 +174,43 @@ void FilterConfigSubscription::onConfigUpdateFailed(Config::ConfigUpdateFailureR
-        config_source, name, factory_context, stat_prefix, *this, subscription_id);
-    subscriptions_.insert({subscription_id, std::weak_ptr<FilterConfigSubscription>(subscription)});
-    return subscription;
-  } else {
-    auto existing = it->second.lock();
-    ASSERT(existing != nullptr,
-           absl::StrCat("Cannot find subscribed filter config resource ", name));
-    return existing;
-  }
 }
 
 FilterConfigProviderPtr FilterConfigProviderManagerImpl::createDynamicFilterConfigProvider(
     const envoy::config::core::v3::ConfigSource& config_source,
     const std::string& filter_config_name, const std::set<std::string>& require_type_urls,
     Server::Configuration::FactoryContext& factory_context, const std::string& stat_prefix,
     bool apply_without_warming) {
-  auto subscription =
-      getSubscription(config_source, filter_config_name, factory_context, stat_prefix);
-  // For warming, wait until the subscription receives the first response to indicate readiness.
-  // Otherwise, mark ready immediately and start the subscription on initialization. A default
-  // config is expected in the latter case.
-  if (!apply_without_warming) {
-    factory_context.initManager().add(subscription->initTarget());
-  }
-  auto provider = std::make_unique<DynamicFilterConfigProviderImpl>(
-      std::move(subscription), require_type_urls, factory_context);
-  // Ensure the subscription starts if it has not already.
-  if (apply_without_warming) {
-    factory_context.initManager().add(provider->init_target_);
+  const std::string provider_identifier =
+      absl::StrCat(MessageUtil::hash(config_source), ".", filter_config_name);
+  auto it = dynamic_filter_config_providers_.find(provider_identifier);
+  if (it == dynamic_filter_config_providers_.end()) {
+    auto subscription = std::make_shared<FilterConfigSubscription>(
+        config_source, filter_config_name, factory_context, stat_prefix);
+    // For warming, wait until the subscription receives the first response to indicate readiness.
+    // Otherwise, mark ready immediately and start the subscription on initialization. A default
+    // config is expected in the latter case.
+    if (!apply_without_warming) {
+      factory_context.initManager().add(subscription->initTarget());
+    }
+    auto provider = std::make_shared<DynamicFilterConfigProviderImpl>(
+        std::move(subscription), require_type_urls, factory_context, *this, provider_identifier);
+    // Ensure the subscription starts if it has not already.
+    if (apply_without_warming) {
+      factory_context.initManager().add(provider->init_target_);
+    }
+    dynamic_filter_config_providers_.insert(
+        {provider_identifier, std::weak_ptr<DynamicFilterConfigProviderImpl>(provider)});
+    return provider;
+  } else {
+    auto existing_provider = it->second.lock();
+    ASSERT(existing_provider != nullptr,
+           absl::StrCat("Cannot find dynamic filter config provider ", filter_config_name));
```

Comment:
This is just going to crash below so I would remove this ASSERT

Reply:
Removed.